### DISCUSSION TITLE
Resource Dictionnary requires adding material resources by hand in order to recognize them

### DISCUSSION THREAD

**Discussion by Lee31416** [unoplatform/uno#11347] [Q&A] [answered]
Hi, 

Is it normal that I need these lines 
![image](https://user-images.githubusercontent.com/72167772/219425611-54811714-e39a-4b93-9661-09d39814be2e.png)
for my resource dictionnary to find the material resources ? These lines were required in a style file (TextBox.xaml) and my resource dictionnary was added after material resources in the App.xaml

**Comment by kazo0** [comment]
Hi @Lee31416,

Those lines should only be required within the MergedDictionaries of your App.xaml. So are you saying you have that in the App.xaml but your custom TextBlock styles still aren't working unless you initialize Material within your TextBlock.xaml file?

Would you be able to provide the contents of your App.xaml so maybe I can spot where the issue is?

**Reply by Lee31416** [reply]
Yes that's exactly what I did. Here's my app.xaml : 

```
<Application x:Class="AppName.App"
			 xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
			 xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
			 xmlns:wasm="http://platform.uno/wasm"
			 xmlns:local="using:AppName"
			 xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
			 xmlns:converters="using:AppName.Converters"
			 xmlns:xamarin="http://company.com/xamarin"
			 mc:Ignorable="wasm xamarin"
			 xmlns:utu="using:Uno.Toolkit.UI"
			 RequestedTheme="Light">

	<Application.Resources>
		<ResourceDictionary>
			<ResourceDictionary.MergedDictionaries>
				<!-- Load WinUI resources -->
				<XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />

				<!-- Load Material Color Palette -->
				<MaterialColors xmlns="using:Uno.Material"
								OverrideSource="ms-appx:///ColorPaletteOverride.xaml" />
				<MaterialFonts xmlns="using:Uno.Material"
							   OverrideSource="ms-appx:///MaterialFontsOverride.xaml" />

				<!-- Load the Material control resources -->
				<MaterialResources xmlns="using:Uno.Material" />
				<ToolkitResources xmlns="using:Uno.Toolkit.UI" />

				<!-- Load Material Toolkit resources -->
				<!--<MaterialToolkitColors xmlns="using:Uno.Toolkit.UI.Material" />-->
				<MaterialToolkitResources xmlns="using:Uno.Toolkit.UI.Material" />

				<!-- Application's control styles -->
				<ResourceDictionary Source="ms-appx:///Styles/Button.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/ComboBox.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/DatePicker.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/FeedView.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/ListView.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/LoadingView.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/MessageListControl.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/NavigationBar.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/SkeletonLoading.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/TabBar.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/TimePicker.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/TripStopControl.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/WebView.xaml" />
				<ResourceDictionary Source="ms-appx:///Styles/TextBox.xaml" />

				<!-- Application's resources -->
				<ResourceDictionary Source="ms-appx:///Converters.xaml" />
				<ResourceDictionary Source="ms-appx:///Icons.xaml" />
			</ResourceDictionary.MergedDictionaries>

			<x:Double x:Key="LandscapeMaxPageWidth">500</x:Double>

			<!-- To override the default FlyoutLightDismissOverlayBackground resource -->
			<!--<SolidColorBrush x:Key="FlyoutLightDismissOverlayBackground"
							 Color="{StaticResource PrimaryBrush}" />-->
			<x:Double x:Key="WideMinWindowWidthh">800</x:Double>

			<Thickness x:Key="ContentDialogPadding">0</Thickness>
		</ResourceDictionary>
	</Application.Resources>
</Application>

```

**Reply by kazo0** [reply]
Very strange, I'm not seeing anything wrong with that App.xaml. Is it possible for you to attach a zip of your project, or if this project is very large, maybe just a zip of a small sample app that reproduces the errors you are experiencing?

**Reply by Lee31416** [reply]
Here u go, 
[PlaceholderBugSample.zip](https://github.com/unoplatform/uno/files/10787153/PlaceholderBugSample.zip)

In this sample if you remove the lines for material colors and material resources in Styles/TextBox.xaml it won't work. Also there's 2 placeholder element in the page because the resource dictionnary I introduced was needed to fix the bug that is featured in this sample. I don't think it's relevant for you but I included it in case it helps figure out the problem. The bug was that the default MaterialOutlinedTextBoxStyle causes a bug in the animation of the placeholder when you start writing in a textbox. Also my app.xaml is different because the one I included was from another project and most of the stuff wasn't needed for the sample.

**Reply by kazo0** [reply]
Hm I'm a little confused, so looking at this sample, it seems that you do not have the Material theme initialized from within the App.xaml. So if you were to remove the `<MaterialColors xmlns="using:Uno.Material" />` and `<MaterialResources xmlns="using:Uno.Material" />` lines from the TextBox.xaml and instead add them to your App.xaml so that it looks like this:

```xml
<Application
    x:Class="teststyle.App"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:teststyle">
	
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />
                <MaterialColors xmlns="using:Uno.Material" />
                <MaterialResources xmlns="using:Uno.Material" />
                <ResourceDictionary Source="ms-appx:///Styles/TextBox.xaml" />
				<!-- Add resource dictionaries here -->
            </ResourceDictionary.MergedDictionaries>
            <!-- Add resources here -->
		</ResourceDictionary>
    </Application.Resources>

</Application>
```

It should work without needing to have the lines in TextBox.xaml

**Reply by Lee31416** [reply]
I tried that while doing the sample and it didn't work if I added the material colors/resources to app.xaml and textbox it would break without giving exceptions. If I added the materials to app xaml only it wasn't finding the resources (the bug).

**Reply by kazo0** [reply]
This is very curious, seems to be some sort of limitation in Windows. I do have example apps where this exact scenario is working fine. Still investigating

**Reply by jeanplevesque** [reply]
@kazo0 What's up with this?

**Reply by jeanplevesque** [reply]
@kazo0 Any update?

**Reply by kazo0** [reply]
@jeanplevesque sorry, investigating again now, will let you know

**Reply by kazo0** [reply, marked as answer]
Okay, so after more investigation, it seems that this is an issue on Windows when you are trying to add a ResourceDictionary containing a SINGLE style within the ResourceDictionary. So, if you were to just add some bogus dummy second Style to that TextBox.xaml, everything should work fine. 

@jeromelaban
Repro app here (notice the crash when trying to run the Windows head, simply uncomment the `MySecondButtonStyle` definition in `Button.xaml` and the app will no longer crash

[UnoApp21.zip](https://github.com/unoplatform/uno/files/11214941/UnoApp21.zip)

I'm assuming maybe Windows is trying to do some fancy inlining optimization if you are trying to include a `ResourceDictionary` that only contains one resource definition??

**Reply by jeanplevesque** [reply]
@Lee31416 Can you try this workaround to see if it resolves the issue?
Basically you need your dictionaries to contain at least **2** resources.
That should allow you keep files like TextBox.xaml clean.

@jeromelaban @kazo0 I'm guessing this is a windows bug that we could report. Where would be the best place to do that?

**Reply by kazo0** [reply]
Yeah I'm still fiddling with it to see what would be a good repro sans uno.material but I think it must be a bug with Windows, right?

**Reply by jeromelaban** [reply]
@kazo0 yeah it may be. We may isolate this in a winappsdk only sample and create an issue on the WinUI 3 repo then.

**Reply by Lee31416** [reply]
Yes I can confirm that it works having 2 styles fixes the problem
